### PR TITLE
Support new client and server API desing

### PR DESCRIPTION
**Pull request by BuddhiWathsala**
## Purpose
- Fixes https://github.com/ballerina-platform/ballerina-standard-library/issues/815
- Fixes https://github.com/ballerina-platform/ballerina-standard-library/issues/795

Design doc:
https://docs.google.com/document/d/1UDPm9bRZ9DS97W5tvZOnolYQsohKYLXLAfrofQJXitI/edit?usp=sharing

## Automation tests
 - Unit tests 
   > Yes

## Test environment
- Java 11
- SLP9
- macOS

### CODE REVIEW

File: grpc-native/src/main/java/org/ballerinalang/net/grpc/listener/ServerCallHandler.java [hunk 1]
```diff
@@ -182,9 +192,10 @@ void onMessageInvoke(ServiceResource resource, Message request, StreamObserver r
         List<Type> signatureParams = resource.getParamTypes();
         Object[] paramValues;
         int i = 0;
-        if (signatureParams.size() >= 1 && CALLER_TYPE.equals(signatureParams.get(0).getName())) {
+        if ((signatureParams.size() >= 1) && (signatureParams.get(0).getTag() == TypeTags.OBJECT_TYPE_TAG) &&
```
Comment (BuddhiWathsala) [inline grpc-native/src/main/java/org/ballerinalang/net/grpc/listener/ServerCallHandler.java:195]: This is the logic I came up with to differentiate 
1. remote functions that return the data and
2. remote functions that use callers

@daneshk please check the validity.

File: grpc-ballerina/tests/01_advanced_type_service.proto [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    string name = 1;
+};
+
+message StockQuotes {
+    repeated StockQuote stock = 1;
+};
+
+message StockNames {
+    repeated string names = 1;
+};
```
Comment (daneshk) [inline grpc-ballerina/tests/01_advanced_type_service.proto:61]:
```suggestion
};

```

File: grpc-ballerina/tests/02_array_field_type_service.proto [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    repeated bool values = 1;
+};
+
+message TestStruct {
+    repeated A values = 1;
+};
+
+message A {
+    string name = 1;
+};
```
Comment (daneshk) [inline grpc-ballerina/tests/02_array_field_type_service.proto:58]:
```suggestion
};

```

File: grpc-ballerina/tests/21_grpc_gzip_encoding_service.proto [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    string description = 3;
+    float price = 4;
+    string destination = 5;
+}
+
+message CombinedShipment {
+    string id = 1;
+    string status = 2;
+    repeated Order ordersList = 3;
+}
```
Comment (daneshk) [inline grpc-ballerina/tests/21_grpc_gzip_encoding_service.proto:40]:
```suggestion
}

```

File: grpc-native/src/main/java/org/ballerinalang/net/grpc/listener/ServerCallHandler.java (hunk 1 above)
Comment (daneshk) [reply grpc-native/src/main/java/org/ballerinalang/net/grpc/listener/ServerCallHandler.java:195]: This is ok. The `Caller` object always comes as the first param. We can validate this in the compiler plugin.

File: grpc-native/src/main/java/org/ballerinalang/net/grpc/stubs/Stub.java [hunk 5]
```diff
@@ -0,0 +1,292 @@
+/*
+ *  Copyright (c) 2018, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (daneshk) [inline grpc-native/src/main/java/org/ballerinalang/net/grpc/stubs/Stub.java:2]:
```suggestion
 *  Copyright (c) 2021, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```

**Review by daneshk — APPROVED** [review]
LGTM

**Pull request merged**